bfdd/bfd.c

```diff
@@ -403,10 +403,21 @@ struct bfd_session *ptm_bfd_sess_find(struct bfd_pkt *cp,
 	if (cp->discrs.remote_discr)
 		return bfd_find_disc(peer, ntohl(cp->discrs.remote_discr));
 
-	/* Search for session without using discriminator. */
-	ifp = if_lookup_by_index(ifindex, vrfid);
-
-	vrf = vrf_lookup_by_id(vrfid);
+	/*
+	 * Search for session without using discriminator.
+	 *
+	 * XXX: we can't trust `vrfid` because the VRF handling is not
+	 * properly implemented. Meanwhile we should use the interface
+	 * VRF to find out which one it belongs.
+	 */
+	ifp = if_lookup_by_index_all_vrf(ifindex);
+	if (ifp == NULL) {
+		if (vrfid != VRF_DEFAULT)
+			vrf = vrf_lookup_by_id(vrfid);
+		else
+			vrf = NULL;
+	} else
+		vrf = vrf_lookup_by_id(ifp->vrf_id);
```

Comment:
Is it not a work around to avoid getting wrong vrfid.

Given the nature of BFD being very aggressive w.r.t interval we might hit may impact scale if we have scaled VRF deployment?

Reply:
> Is it not a work around to avoid getting wrong vrfid.

Yes.

> Given the nature of BFD being very aggressive w.r.t interval we might hit may impact scale if we have scaled VRF deployment?

I think we might have worst bottlenecks, but, sure, it will have an impact on system with more interfaces and/or VRFs. The complexity will jump from `O(log m)` to `O(n log m)` where `n` is the number of VRFs and `m` is the number of interfaces (actually a bit slower than that, because `if_lookup_by_ifindex`[1] causes another VRF lookup).

At the moment we need to fix a even bigger problem which is described at #5390 before we get the performance improvement here. I think it would be more productive to profile the code and get the exact bottlenecks before spending time here.

[1]: https://github.com/FRRouting/frr/blob/stable/7.2/lib/if.c#L273

Reply:
@rzalamena - shouldn't we be fixing `vrf_bind` which is so clearly broken?

Reply:
@qlyoung I don't know how to fix it, but I've made a suggestion in the issue about this.

Meanwhile this workaround fixes the problem.